var/spack/repos/builtin/packages/netcdf-c/package.py

```diff
@@ -102,7 +102,8 @@ def url_for_version(self, version):
 
     # High-level API of HDF5 1.8.9 or later is required for netCDF-4 support:
     # http://www.unidata.ucar.edu/software/netcdf/docs/getting_and_building_netcdf.html
-    depends_on('hdf5@1.8.9:+hl')
+    depends_on('hdf5@1.8.9:+hl api=v18', when='@4.8.0:')
+    depends_on('hdf5@1.8.9:+hl', when='@:4.8.0')
```

Comment:
Just a nitpick. Since the upper directive is a constraint on the lower I think you can omit the `when` argument:
```suggestion
    depends_on('hdf5@1.8.9:+hl')
```